### PR TITLE
Translations 2.0

### PR DESCRIPTION
**Pull request by tumbledwyer**
## Description

Update translations to be a file per language for ease of updates.

Changes:
- Core now publishes translation as part of toolkit
- Core translations are stored in core-xx.csv per language
- Country config translations are stored in country-xx.csv per language
- Country config validates translations from core and from itself.
- `yarn validate:translations` checks for all missing, unused or blank translation
    - Script can fix or printout errors
    - Is used in the github action
    - Shows errors at app startup

## Core Dev Workflow
- Make a change in core that adds a new translation key
- Build the npm toolkit in github actions
- Create a countryconfig branch with the same name
- Set the toolkit version in `package.json`
- Add translation key to `core-en.csv` and `core-fr.csv`
    - Do this manually
    - or run `yarn validate:translations` in countryconfig and the script will show that you have a missing translation and offer to add the key for you (you will still need to manually enter the french translation)

### When upgrading to v2.0 (included in [translations/README.md](https://github.com/opencrvs/opencrvs-countryconfig/pull/1353/changes#diff-3ad3dad5ac9d82c95a1d54b83d0c35a8b5ef5d6faf06e36e9e424dde33d6b399))
- Merge in v2.0 country config into your country
- Run `yarn validate:translations`
- Select Migrate
- Script merges `client.csv` and `login.csv` and generates `core-xx.csv` and `country-xx.csv` for all languages with only the valid keys in both core and country config
- If there are blank keys (keys exist but have no translation) the script can print them out so they can be translated

### When updating to a minor/patch release
- Merge in country config
- Run `yarn validate:translations`
- If there are unused (stale) keys, you will be given the option to remove them
- If there are missing (new) keys, you will be given an option add them to `core-xx.csv`
    - Once added you will get the error that they are blank in your non-english `core-xx.csv` and need to be translated

### When adding fields to a form
- Add fields and run `yarn validate:translations`
- You should get a missing keys error with the option to add keys all of your `country-xx.csv` files
- Once added, you'll get a blank keys error and can print the keys requiring translation in your other language files

### From country config
Github actions runs `yarn validate:translations:ci` which validates against the current toolkit version. 
If there are any errors the job fails.

### From core
GIthub actions checks out core and country config with the same branch name.
Generates translations in core (`core-en.json`) and runs yarn install on country config to pull the toolkit with translations.
If the translation file is not the same on both sides, which means you didn't update the toolkit in cc, the job fails.
Runs`yarn validate:translations:ci` in country config and fails if there are errors.

## Checklist

- [ ] I have linked the correct Github issue under "Development"
- [ ] I have tested the changes locally, and written appropriate tests
- [ ] I have tested beyond the happy path (e.g. edge cases, failure paths)
- [ ] I have updated the changelog with this change (if applicable)
- [ ] I have updated the GitHub issue status accordingly

### CODE REVIEW

File: src/api/content/service.ts [hunk 1]
```diff
@@ -28,11 +29,35 @@ export type CSVRow = { id: string; description: string } & Record<
...
+    : []
+
+  if (perLanguageFiles.length > 0) {
+    return mergePerLanguageFiles(perLanguageFiles)
+  }
+
   const csvData = await readCSVToJSON<CSVRow[]>(
-    join('src/translations/', `${application}.csv`)
+    join(translationsDir, `${application}.csv`)
   )
```
Comment (Nil20) [inline src/api/content/service.ts]: Since `client.csv` and `login.csv` are deleted in this PR, so this fallback should throw for both getLanguages('client') and getLanguages('login').                                                     
                                                                                                                                                                                              
  This is masked locally if you have cached offline data in the browser — client translations are hydrated from IndexedDB on startup (GET_OFFLINE_DATA_SUCCESS in the offline reducer), so the failed network call may go unnoticed. Login has no such cache and will visibly break.                                                                                                        
                              
  To reproduce: clear browser storage and reload. Neither client nor login translations will load.                                                                                            
                              
  service.ts needs logic to merge core-*.csv + country-*.csv into a combined response when the legacy ${application}.csv files no longer exist.

File: src/validate-translations.ts [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+      return
+    }
+    try {
+      // eslint-disable-next-line no-new-func
+      matches.push(new Function(`return (${node.getText(sourceFile)});`)())
+    } catch {
+      // Dynamic identifier — skip
+    }
+    ts.forEachChild(node, visit)
+  }
```
Comment (Nil20) [inline src/validate-translations.ts:145]: Same new Function() fragility as in the core PR script — TypeScript-specific syntax in any descriptor will silently drop the key. Same AST string literal fix could be applied here. Also worth logging skipped descriptors (with file path and a snippet of the object text) so one can see what was intentionally excluded vs silently broken.                                     
                                                                                                                                                                                              
  Less critical here than in core since country config source tends to be simpler and less likely to have complex TypeScript syntax in message descriptors — but might be worth fixing for consistency.

File: src/translations/README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
+## Managing translations
+
+### Translation changes in 2.0
+- Removed `client.csv` and `login.csv` which contained translations per service
+- Created a `core-xx.csv` file per language containing core translations
+- Created a `country-xx.csv` file per language containing country config translations
+- Include core translations in OpenCRVS toolkit
+- TODO maybe add something about vaildation here
```
Comment (Nil20) [inline src/translations/README.md:8]:
```suggestion
- TODO maybe add something about validation here
```

**Review by Nil20 — COMMENTED** [review]
Nice architectural improvement! 🙌

File: src/api/content/service.ts (hunk 1 above)
Comment (tumbledwyer) [reply src/api/content/service.ts]: Thanks, I missed that. Fixed now.

File: src/validate-translations.ts (hunk 2 above)
Comment (tumbledwyer) [reply src/validate-translations.ts:145]: Done

File: src/validate-translations.ts [hunk 4]
```diff
@@ -0,0 +1,699 @@
...
+        descriptionNode && ts.isStringLiteral(descriptionNode)
+          ? descriptionNode.text
+          : undefined
+
+      if (ts.isStringLiteral(idNode) && ts.isStringLiteral(msgNode)) {
+        matches.push({
+          id: idNode.text,
+          defaultMessage: msgNode.text,
+          description: description
+        })
```
Comment (Nil20) [inline src/validate-translations.ts]: https://github.com/opencrvs/opencrvs-core/pull/12209#discussion_r3232151838 - same as core! could `isNoSubstitutionTemplateLiteral` be considered for validation?

File: src/validate-translations.ts [hunk 5]
```diff
@@ -0,0 +1,699 @@
...
+      } else {
+        console.warn(
+          `Skipping non-literal descriptor : ${node.getText(sourceFile)}`
```
Comment (Nil20) [inline src/validate-translations.ts]:
```suggestion
          `Skipping non-literal descriptor : ${node.getText(sourceFile).slice(0, 80)}`
```

File: src/validate-translations.ts [hunk 6]
```diff
@@ -0,0 +1,699 @@
...
+        )
+      }
+    } catch {
+      console.warn(`Skipping dynamic descriptor : ${node.getText(sourceFile)}`)
```
Comment (Nil20) [inline src/validate-translations.ts]:
```suggestion
      console.warn(`Skipping dynamic descriptor : ${node.getText(sourceFile).slice(0, 80)}`)
```

File: src/validate-translations.ts [hunk 7]
```diff
@@ -0,0 +1,699 @@
...
+    console.warn(
+      '\nLegacy translations detected. Select "Migrate translations" to update them to 2.0.\n'
+    )
+  }
+
+  // Interactive loop (warnOnly === false)
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    languages = detectLanguages(translationsDir)
+    const currentIssues = await validateAll(
```
Comment (Nil20) [inline src/validate-translations.ts:621]: `detectLanguages` returns only languages that have a core-{lang}.csv file, right? `validateAll` then loops over that list. So if `country-fr.csv` exists but `core-fr.csv` does not, French translations woulnd't the loop at all. `country-fr.csv` is never opened and validated either, and no warning is printed. From the tool's perspective French doesn't exist.

The fix can be like this - after `detectLanguages` returns, scan for any `country-{lang}.csv` whose lang is not in the detected set and warn.

File: src/translations/core-fr.csv
Comment (Nil20) [inline src/translations/core-fr.csv:1]: `core-en` and `core-fr` has different numbers of translations. Shouldn't they ideally be the same in number? Maybe make sure, they are not faulty and the script cover the cases.

File: src/validate-translations.ts (hunk 4 above)
Comment (tumbledwyer) [reply src/validate-translations.ts]: Done

File: src/validate-translations.ts (hunk 5 above)
Comment (tumbledwyer) [reply src/validate-translations.ts]: Done

File: src/validate-translations.ts (hunk 6 above)
Comment (tumbledwyer) [reply src/validate-translations.ts]: Done

File: src/validate-translations.ts (hunk 7 above)
Comment (tumbledwyer) [reply src/validate-translations.ts:621]: Done

File: src/translations/core-fr.csv
Comment (tumbledwyer) [reply src/translations/core-fr.csv:1]: The number of translations was actually the same. But for some reason, the formatting was different and the en file had line breaks so the total lines was different.
i manually edited it so it looks the same now.

**Review by Nil20 — APPROVED** [review]
